File: leetcode/0987-reveal-cards-in-increasing-order/solution.go
func deckRevealedIncreasing(deck []int) []int {
    slices.Sort(deck)
    ret := make([]int, len(deck))
    q := make([]int, len(deck))
    for i := range q {
        q[i] = i
    }
    take := true
    for len(q) != 0 {
        i := q[0]
        q = q[1:]
        if take {
            x := deck[0]
            deck = deck[1:]
            ret[i] = x
        } else {
            q = append(q, i)
        }
        take = !take
    }
    return ret
}

File: leetcode/0987-reveal-cards-in-increasing-order/solution.rs
impl Solution {
    pub fn deck_revealed_increasing(mut deck: Vec<i32>) -> Vec<i32> {
        use std::collections::VecDeque;
        deck.sort_by(|a,b|b.cmp(&a));
        let n = deck.len();
        let mut ret = vec![0;n];
        let mut q: VecDeque<usize> = (0..n).collect();
        let mut take = true;
        while let Some(i) = q.pop_front() {
            if take {
                if let Some(x) = deck.pop() {
                    ret[i] = x;
                }
            } else {
                q.push_back(i);
            }
            take = !take;
        }
        ret
    }
}
